apm-agent/src/test/java/co/elastic/apm/agent/PackagingTest.java

```diff
@@ -18,83 +18,300 @@
+            this.basePackage = getBasePackage(rootFolder);
+
+            // verify module invariants & store effective groupID
+            String groupId = mavenModel.getGroupId();
+            if (groupId == null) {
+                assertThat(mavenModel.getParent()).isNotNull();
+                groupId = mavenModel.getParent().getGroupId();
+            }
+            this.mavenGroupId = groupId;
+            assertThat(groupId)
+                .isEqualTo("co.elastic.apm");
+        }
+
+        public boolean isPlugin() {
+            try {
+                Path servicesFolder = getServicesFolder();
+                return servicesFolder != null && Files.list(servicesFolder).count() > 0;
+            } catch (IOException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        public boolean hasCode() {
+            return basePackage != null;
+        }
+
+        public Collection<String> getClassloaderRootPackages() {
+            Path servicesFolder = getServicesFolder();
+            if (servicesFolder != null) {
+                Path customizerService = servicesFolder.resolve(PluginClassLoaderRootPackageCustomizer.class.getName());
+                if (Files.exists(customizerService)) {
+                    try {
+                        List<String> lines = Files.readAllLines(customizerService);
+                        assertThat(lines).describedAs("exactly one root package customizer expected in file %s", customizerService).hasSize(1);
+                        Class<?> customizerType = Class.forName(lines.get(0), true, PackagingTest.class.getClassLoader());
+                        PluginClassLoaderRootPackageCustomizer customizer = (PluginClassLoaderRootPackageCustomizer) customizerType.getConstructor().newInstance();
+                        return customizer.pluginClassLoaderRootPackages();
+                    } catch (Exception e) {
+                        throw new IllegalStateException(e);
+                    }
+                }
+            }
+            return Collections.singleton(basePackage);
+        }
+
+        @Nullable
+        private Path getServicesFolder() {
+            Path servicesFolder = rootFolder.resolve(Path.of("src", "main", "resources", "META-INF", "services"));
+            return Files.isDirectory(servicesFolder) ? servicesFolder : null;
+        }
+
+        public Set<String> getInternalDependencies() {
+            return mavenModel.getDependencies()
+                .stream()
+                .filter(d -> getGroupId(d).equals(mavenGroupId))
+                .filter(d -> d.getScope() == null || !d.getScope().equals("test"))
                 .map(Dependency::getArtifactId)
                 .collect(Collectors.toSet());
+        }
 
-            assertThat(dependenciesArtifacts).describedAs(createDescription(plugins)).containsAll(plugins);
+        public boolean hasDependencies() {
+            return !mavenModel.getDependencies().isEmpty();
+        }
```

Comment:
[minor] just an idea to allow for the module having test dependencies

```suggestion
        public boolean hasNonTestDependencies() {
            return mavenModel.getDependencies().stream().anyMatch(d -> !d.getScope().equals("test"));
        }
```